libmamba/include/mamba/download/request.hpp

```diff
@@ -0,0 +1,179 @@
+        Content content = {};
+        TransferData transfer = {};
+        std::string cache_control = "";
+        std::string etag = "";
+        std::string last_modified = "";
+        std::size_t attempt_number = std::size_t(1);
+    };
+
+    struct Error
+    {
+        std::string message = "";
+        std::optional<std::size_t> retry_wait_seconds = std::nullopt;
+        std::optional<TransferData> transfer = std::nullopt;
+        std::size_t attempt_number = std::size_t(1);
+    };
+
+    using Result = tl::expected<Success, Error>;
+    using MultiResult = std::vector<Result>;
+
+    /*****************************
+     * Download event structures *
+     *****************************/
+
+    struct Progress
+    {
+        std::size_t downloaded_size = 0;
+        std::size_t total_to_download = 0;
+        std::size_t speed_Bps = 0;
+    };
+
+    using Event = std::variant<Progress, Error, Success>;
+
+    /*******************************
+     * Download request structures *
+     *******************************/
+
+    struct RequestBase
+    {
+        using progress_callback_t = std::function<void(const Event&)>;
+
+        // TODO: remove these functions when we plug a library with continuation
+        using on_success_callback_t = std::function<expected_t<void>(const Success&)>;
+        using on_failure_callback_t = std::function<void(const Error&)>;
+
+        std::string name;
+        // If filename is not initialized, the data will be downloaded
+        // to an internal in-memory buffer.
+        std::optional<std::string> filename;
+        bool check_only;
+        bool ignore_failure;
+        std::optional<std::size_t> expected_size = std::nullopt;
+        std::optional<std::string> etag = std::nullopt;
+        std::optional<std::string> last_modified = std::nullopt;
+
+        std::optional<progress_callback_t> progress = std::nullopt;
+        std::optional<on_success_callback_t> on_success = std::nullopt;
+        std::optional<on_failure_callback_t> on_failure = std::nullopt;
+
+    protected:
+
+        RequestBase(
+            std::string_view lname,
+            std::optional<std::string> lfilename,
+            bool lcheck_only,
+            bool lignore_failure
+        );
+
+        ~RequestBase() = default;
+        RequestBase(const RequestBase&) = default;
+        RequestBase& operator=(const RequestBase&) = default;
+        RequestBase(RequestBase&&) = default;
+        RequestBase& operator=(RequestBase&&) = default;
+    };
+
+    // This class is used to create strong alias on
+    // string_view. This helps to avoid error-prone
+    // calls to functionsthat accept many arguments
+    // of the same type
+    template <int I>
+    class string_view_alias
+    {
+    public:
+
+        explicit string_view_alias(std::string_view s);
+        operator std::string_view() const;
+
+    private:
+
+        std::string_view m_wrapped;
+    };
+
+    using MirrorName = string_view_alias<0>;
```

Comment:
Why not?
```suggestion
struct MirrorName
{
    std::string_view name;
}
```

Reply:
Because initially I also wanted to have aliases for the `url_path` and why not the`name` fields of the request, and I found this solution more elegant than repeating the same code for wrapping the same type. I had some problem with wrapping the url that I postponed to a future PR. Ideally, this structure should be moved to some util files, and accept an additional template argument (for the type it should alias). But I agree it may seem overkill actually.

A minor difference is that the structure forces a {} initialization while the alias lets the choice.